core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -126,9 +127,15 @@ private[kafka] object LogValidator extends Logging {
     for (batch <- records.batches.asScala) {
       validateBatch(batch, isFromClient, toMagicValue)
 
+      var recordCount = 0
       for (record <- batch.asScala) {
         validateRecord(batch, record, now, timestampType, timestampDiffMaxMs, compactedTopic)
         builder.appendWithOffset(offsetCounter.getAndIncrement(), record)
+        recordCount += 1
+      }
+      if (batch.isInstanceOf[DefaultRecordBatch]) {
+        if(batch.asInstanceOf[DefaultRecordBatch].count() != recordCount)
```

Comment:
Instead of doing this we can add a `countOrNull` method that returns `null` for the old message formats. It would also be good if we can share the logic a bit more.